R/modify.R

```diff
@@ -195,76 +136,59 @@ modify_at <- function(.x, .at, .f, ...) {
 #' @export
 modify_at.default <- function(.x, .at, .f, ...) {
   where <- where_at(.x, .at)
-  modify_if(.x, where, .f, ...)
-}
-#' @export
-modify_at.integer <- function(.x, .at, .f, ...) {
-  where <- where_at(.x, .at)
-  .x[where] <- map_int(.x[where], .f, ...)
-  .x
-}
-#' @export
-modify_at.double <- function(.x, .at, .f, ...) {
-  where <- where_at(.x, .at)
-  .x[where] <- map_dbl(.x[where], .f, ...)
-  .x
-}
-#' @export
-modify_at.character <- function(.x, .at, .f, ...) {
-  where <- where_at(.x, .at)
-  .x[where] <- map_chr(.x[where], .f, ...)
-  .x
-}
-#' @export
-modify_at.logical <- function(.x, .at, .f, ...) {
-  where <- where_at(.x, .at)
-  .x[where] <- map_lgl(.x[where], .f, ...)
-  .x
+  modify_where(.x, where, .f, ...)
 }
 
-# modify2 -----------------------------------------------------------------
-
 #' @rdname modify
 #' @export
 modify2 <- function(.x, .y, .f, ...) {
   UseMethod("modify2")
 }
 #' @export
 modify2.default <- function(.x, .y, .f, ...) {
-  modify2_base(map2, .x, .y, .f, ...)
-}
-# TODO: Improve genericity (see above)
-#' @export
-modify2.integer  <- function(.x, .y, .f, ...) {
-  modify2_base(map2_int, .x, .y, .f, ...)
-}
-#' @export
-modify2.double  <- function(.x, .y, .f, ...) {
-  modify2_base(map2_dbl, .x, .y, .f, ...)
-}
-#' @export
-modify2.character  <- function(.x, .y, .f, ...) {
-  modify2_base(map2_chr, .x, .y, .f, ...)
-}
-#' @export
-modify2.logical  <- function(.x, .y, .f, ...) {
-  modify2_base(map2_lgl, .x, .y, .f, ...)
-}
-
-modify2_base <- function(mapper, .x, .y, .f, ...) {
   .f <- as_mapper(.f, ...)
-  out <- mapper(.x, .y, .f, ...)
 
-  # if .x got recycled by map2
-  if (length(out) > length(.x)) {
-    .x <- .x[rep(1L, length(out))]
+  if (vec_is_list(.x) || is.data.frame(.x)) {
+    out <- map2(vec_proxy(.x), .y, .f, ...)
+    vec_restore(out, .x)
+  } else if (vec_is(.x)) {
```

Comment:
This is simultaneously technically correct and also makes me sad that we generate the duplicated column names

``` r
modify2(data.frame(x = 1), 1:3, \(x, y) x)
#>   x x x
#> 1 1 1 1
```